### ISSUE TITLE
DNSCrypt upstream does not retry over TCP when it gets a truncated response

### ISSUE THREAD

**Issue by evilvibes** [AdguardTeam/AdGuardHome#524]
Seeing odd behaviour when using a DNScrypt Stamp. Youtube will not load and Tidal music will not play any music (with in the app). When using Cloudflare DoH or DoT it will when Cloudflare works.

Half the time Cloudflare will not work at all.

Using Adguard Home Windows 10.

**Comment by ameshkov** [comment]
I am quite sure this is actually a duplicate of #520

**Comment by ameshkov** [comment]
Could you please run AG Home with `-v` argument and capture the verbose output? I'd like to see what's wrong there.

**Comment by evilvibes** [comment]
The only way I can get cloudflare to work is tls://1.1.1.1 or just 1.1.1.1 it seems all other ways will not work. My goal is to use my DNScrypt servers for upstream but when using them as I said before some things will not open/load. When I get some free time i'll look more into it for you.

**Comment by ameshkov** [comment]
Anyways, if you run AGH like this `AdGuardHome.exe -v` it will print verbose log to the stdout. It's interesting to see what's printed there when something begins to misbehave.

**Comment by evilvibes** [comment]
Seems to be working now. The issue seem to be when using my own dnscrypt servers when I used lets say opendns dnscrypt server everything would work right.

My two servers are:
sdns://AQcAAAAAAAAADjE0OS4yNDguNTEuMjEyIJqd7T8sjWmQU5pdBaP5FF2vX0xrz561gdeGrur3jCHfHTIuZG5zY3J5cHQtY2VydC5ldmlsdmliZXMuY29t

sdns://AQcAAAAAAAAADjEwNC4xNTYuMjUyLjE4ILNYRzmUAe4JezM_eiaokRdEOF7hA_jTl65AAyNMlNYQHTIuZG5zY3J5cHQtY2VydC5ldmlsdmliZXMuY29t

Try testing them from your end see if youtube will open.

**Comment by ameshkov** [comment]
Yeah, I see what's going on, thank you!

The response over UDP is truncated and to get the proper response, we should retry over TCP:
```
$ nslookup www.youtube.com 10.7.113.140
;; Truncated, retrying in TCP mode.
;; Connection to 10.7.113.140#53(10.7.113.140) for www.youtube.com failed: connection refused.
```

**Comment by ameshkov** [comment]
@hmage the issue is resolved in dnsproxy v0.9.11: https://github.com/AdguardTeam/dnsproxy/releases/tag/v0.9.11

**Comment by ameshkov** [comment]
Fixed
